### ISSUE TITLE
load 72B teacher model

### ISSUE THREAD

**Issue by ypw-lbj** [songmzhang/DSKD#13]
For a larger teacher model, such as 72B, how should I modify the code of this framework? Please give me some guidance. 
Thank you.

**Comment by songmzhang** [comment]
Sorry, due to the device limitation, we haven't tried to load a 72B teacher model in our experiments. Will you encounter any issue if you just modify the TEACHER_MODEL_PATH to the path of the 72B teacher model?

**Comment by ypw-lbj** [comment]
> Sorry, due to the device limitation, we haven't tried to load a 72B teacher model in our experiments. Will you encounter any issue if you just modify the TEACHER_MODEL_PATH to the path of the 72B teacher model?

I use 8 GPUs. If I use 72B_path, I will get OOM problem. So I change GPUS_PER_NODE=1, and then change AutoModelForCausalLM.from_pretrained(
device_map=“auto”, }, but there will be conflict in deepspeed.initialize and OOM problem will occur.
I hope this repo can support larger models and will have greater influence.
Thank you

**Comment by songmzhang** [comment]
> > Sorry, due to the device limitation, we haven't tried to load a 72B teacher model in our experiments. Will you encounter any issue if you just modify the TEACHER_MODEL_PATH to the path of the 72B teacher model?
> 
> I use 8 GPUs. If I use 72B_path, I will get OOM problem. So I change GPUS_PER_NODE=1, and then change AutoModelForCausalLM.from_pretrained( device_map=“auto”, }, but there will be conflict in deepspeed.initialize and OOM problem will occur. I hope this repo can support larger models and will have greater influence. Thank you

Thanks for your suggestion! We will work on it in recent weeks. BTW, I think a more efficient way to use such a large teacher model is to pre-compute the logits of the teacher and save them before KD. Anyway, we will try to fix this issue as soon as possible.

**Comment by botox-100** [comment]
> > > Sorry, due to the device limitation, we haven't tried to load a 72B teacher model in our experiments. Will you encounter any issue if you just modify the TEACHER_MODEL_PATH to the path of the 72B teacher model?
> > 
> > 
> > I use 8 GPUs. If I use 72B_path, I will get OOM problem. So I change GPUS_PER_NODE=1, and then change AutoModelForCausalLM.from_pretrained( device_map=“auto”, }, but there will be conflict in deepspeed.initialize and OOM problem will occur. I hope this repo can support larger models and will have greater influence. Thank you
> 
> Thanks for your suggestion! We will work on it in recent weeks. BTW, I think a more efficient way to use such a large teacher model is to pre-compute the logits of the teacher and save them before KD. Anyway, we will try to fix this issue as soon as possible.

Hi, is there any news on this topic? I run into the same issue and did not find an (easy) way to workaround. 
Thanks for your effort, Thomas

**Comment by songmzhang** [comment]
> > > > Sorry, due to the device limitation, we haven't tried to load a 72B teacher model in our experiments. Will you encounter any issue if you just modify the TEACHER_MODEL_PATH to the path of the 72B teacher model?
> > > 
> > > 
> > > I use 8 GPUs. If I use 72B_path, I will get OOM problem. So I change GPUS_PER_NODE=1, and then change AutoModelForCausalLM.from_pretrained( device_map=“auto”, }, but there will be conflict in deepspeed.initialize and OOM problem will occur. I hope this repo can support larger models and will have greater influence. Thank you
> > 
> > 
> > Thanks for your suggestion! We will work on it in recent weeks. BTW, I think a more efficient way to use such a large teacher model is to pre-compute the logits of the teacher and save them before KD. Anyway, we will try to fix this issue as soon as possible.
> 
> Hi, is there any news on this topic? I run into the same issue and did not find an (easy) way to workaround. Thanks for your effort, Thomas

Sorry for delaying this issue for a while since I've been busy with other work. I will first try to fix the issue when `device_map="auto"` and see whether we can load the 72b model with pipeline parallelism. Theoretically, if we want to load a 72B model in bf16 during KD, at least we need to leave 72*2=144G memory totally on all GPUs (e.g., >18G on each GPU if num_gpus=8). It poses high requirements on the devices. Here are some potential solutions:
- Load a 72b model with pipeline/tensor parallelism to split it into multiple GPUs;
- Load the INT8/4 quantized version to reduce memory cost (not sure how it will affect the performance);
- Expand to multiple nodes (machines) and use one for specifically loading the 72b model.
- Pre-compute the (top-k) logits from the 72B model on the training data and save them together, then KD only needs these logits and does not require to load the 72b model.

**Comment by songmzhang** [comment]
Hi guys! We have updated our code which supports knowledge distillation from Qwen2.5-72B to Qwen2.5-1.5B using deepspeed ZeRO-3. You can now try it with `--deepspeed_config ./config/deepspeed/ds_config_zero3_bf16/fp16.json` (Note that the distillation from a 72B model to a 1.5B model is quite costly and we use 8*A100 40G to barely achieve it. Please make sure you have enough GPU memory before you try).
